plugins/blockvolume/hostvol/utils.go

```diff
@@ -15,34 +15,70 @@ import (
 	"github.com/gluster/glusterd2/glusterd2/volume"
 	"github.com/gluster/glusterd2/pkg/api"
 	"github.com/gluster/glusterd2/pkg/size"
-
 	log "github.com/sirupsen/logrus"
 )
 
+func filterSize(volinfo *volume.Volinfo, size uint64) bool {
+	availableSize, found := volinfo.Metadata[volume.BlockHostingAvailableSize]
+	if !found {
+		return false
+	}
+	if availableSizeInBytes, err := strconv.ParseUint(availableSize, 10, 64); err == nil && availableSizeInBytes > size {
+		return true
+	}
+
+	return false
+}
+
+func applyArbiter(volinfo *volume.Volinfo, arbPath string) bool {
+	if arbPath == "" {
+		return true
+	}
+	_, exists := volinfo.Options["cluster/replicate.thin-arbiter"]
+	if exists {
+		return true
+	}
+	return false
+}
+
+func applyShard(volinfo *volume.Volinfo, size uint64) bool {
+	if size == 0 {
+		return true
+	}
+	_, exists := volinfo.Options["features/shard"]
+	if exists {
+		//Checking for shard size may not be required
+		return true
+	}
+	return false
+}
+
 // BlockSizeFilter returns a volume Filter, which will filter out volumes
 // haing block-hosting-available-size greater than give size.
-func BlockSizeFilter(size uint64) volume.Filter {
+func HostOptnsFilter(size uint64, h *HostingVolumeOptions) volume.Filter {
 	return func(volinfos []*volume.Volinfo) []*volume.Volinfo {
 		var volumes []*volume.Volinfo
 
 		for _, volinfo := range volinfos {
-			availableSize, found := volinfo.Metadata[volume.BlockHostingAvailableSize]
-			if !found {
+			if !filterSize(volinfo, size) {
 				continue
 			}
-
-			if availableSizeInBytes, err := strconv.ParseUint(availableSize, 10, 64); err == nil && availableSizeInBytes > size {
-				volumes = append(volumes, volinfo)
+			if !applyArbiter(volinfo, h.ThinArbPath) {
+				continue
+			}
+			if !applyShard(volinfo, h.ShardSize) {
+				continue
 			}
+			volumes = append(volumes, volinfo)
 		}
 		return volumes
 	}
 }
 
 // GetExistingBlockHostingVolume returns a existing volume which is suitable for hosting a gluster-block
-func GetExistingBlockHostingVolume(size uint64) (*volume.Volinfo, error) {
+func GetExistingBlockHostingVolume(size uint64, h *HostingVolumeOptions) (*volume.Volinfo, error) {
 	var (
-		filters     = []volume.Filter{volume.FilterBlockHostedVolumes, BlockSizeFilter(size)}
+		filters     = []volume.Filter{volume.FilterBlockHostedVolumes, HostOptnsFilter(size, h)}
```

Comment:
use can use filters like []volume.Filter{volume.FilterBlockHostedVolumes, BlockSizeFilter(size),FilterArbiter,FilterShard}.
FilterArbiter,FilterShard need to be implemented

Reply:
I thought of the same, but if i were to apply filter separately then each filter has to loop though the list of all volumes right? So wrote it like this, for each volume apply all filters required. If the number of volumes are huge, this looping of volumes for each filter adds up redundancy? But yeah block hosting volumes are generally less in number, hence not much difference may be. Thought?

Reply:
right, if the number of volumes will be more then it will be a problem, okay then we can go with the current implementation.

Reply:
Done